boiling detection

mariamedi opened the pull request:
## Changes
- Constructed a boiling detector class
- Split an input frame into quadrants
- Compared last and current quadrants for SSIM
- Found median SSIM every X recordings
- Detected boiling by calculating if average was under a thresh

mariamedi committed:
constructed boil detector class

mariamedi committed:
cleaned up some comments, minor fixes

bhylak commented:
Hey,

Congrats on your first PR! Here's some things to fix:

# Changes
- Use headings!
- Include Pictures
- Document all of the new functionality (just in bullet points!)

# Known Issues
- then talk about any known issues with the code that *could* be a problem down the line.

Then ask your questions on a comment in the PR.

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+                i = i + 1
+
+        # sets appropriate stove burners to boiling
+        if burners[0].boiling and stove.upper_left.temp >= self.temp:
```
rename self.temp to "MINIMUM_BOILING_TEMP"

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+            self.first_quads = self.__split_frame(self.first_image)
+
+        # if the number of frames that have been processed are desired amount, proceed
+        elif self.frame_count == self.count:
```
rename self.count to something more descriptive

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+                self.second_quads = self.__split_frame(self.second_image)
+            # second image does exist, must shift images
+            else:
+                self.first_image = self.second_image
```
first image, second image. Why isn't there a third image? :) Is there a better name for these? What does it do?

benhylak replied:
once it's streaming -- current image, last image

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+            self.second_quads = self.__split_frame(self.second_image)
+
+            # Detects boiling on active burners
+            i = 0
```
*i* don't like the name of this variable :)

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+            for burner in burners:
+                # if there is a pot on burner
+                if burner.pot_detected:
+                    mask = self.masks[i]
```
so this means that the mask is never changed -- is that correct?

maybe it would be better to change the mask if the difference is above some sort of threshold -- does that make sense?

mariamedi replied:
right now the mask is changed whenever a new pot is detected on the stove. I'm a little confused on how it would be better to change the mask above a certain threshold?

benhylak replied:
If the pot is placed on the stove and then moved. At the least, you'll probably want a buffer -- like, calculate the threshold after 15 seconds of it being on the stove.

Otherwise, it's going to calculate that threshold the second it comes into the camera view, which will probably be as it's being placed.

mariamedi replied:
Believe I fixed this by breaking out of the update function when the calculated SSIM is below MASK_THRESH

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+                if burner.pot_detected:
+                    mask = self.masks[i]
+                    score = self.___get_score(mask, self.first_quads[i], self.second_quads[i])
+                    self.median_SSIM[i].append(score)
```
it's not appending to the median, but rather an array of potential median values, right?

mariamedi replied:
yes

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+                    self.median_SSIM[i].append(score)
+
+                    # median is calculated when enough SSIMs have been recorded
+                    if len(self.median_SSIM[i]) == self.SSIM_count:
```
talk to mona -- she's gonna implement a StreamingAverage class that will help like this.

This means we can update more often while also taking averages over a window.

It's really just a queue. Every loop it pops an element (if the length is > then the window) and adds a new reading. Every loop, it also takes an average across the whole queue.

melokda replied:
Yes, it will be included in the main file so I'll let you know once I implement it and test it and then you can use it.

mariamedi replied:
Included this in the most recent commits

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+        else:
+            return False
+
+    def __burners(self,stove):
```
this should be added to the stove as a function like "get_burners()"

benhylak replied:
and then refactored through the rest.

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+                i = i + 1
+
+        # sets appropriate stove burners to boiling
+        if burners[0].boiling and stove.upper_left.temp >= self.temp:
```
I think this can be refactored to foreach(burner in stove.get_burners()) (once you implement a get_burners function, like mentioned above)

mariamedi replied:
Question: if I have the following code,

        for burner, burner_boil in zip(burners,self.boiling):
            if burner_boil and burner.temp >= self.MIN_BOILING_TEMP:
                burner.boiling = True 
            else:
                burner.boiling = False 

will the "burner.boiling" change the actual value of the stove burner? not just the "burner" object being used to iterate?

benhylak reviewed (changes requested):
Good work!

melokda commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -0,0 +1,240 @@
+class BoilDetector:
+
+    def __init__self(self):
+        self.temp = 170 # temp considered to be boiling
```
We need to verify this with actual experimentation, but it will probably be way less than this.

melokda reviewed (commented):
I know it's difficult for you to test it because you're away, but we really need to test it live soon. I would be happy to do a live test when I come back to Worcester on Thursday once you implement these changes. I can also help with finding the right temperature for boiling consideration.

mariamedi committed:
modified variable names, refactored get_burners

mariamedi committed:
added queues

mariamedi commented:
## Changes
- Modified variable names
- Added queues for median SSIM and boiling

## Known Issues
- Still only update mask when a new pot is detected, may require modification
- Boiling min temp must be calculated/found

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -2,27 +2,34 @@
+                            queue.Queue(maxsize=self.SSIM_count),
+                            queue.Queue(maxsize=self.SSIM_count),
+                            queue.Queue(maxsize=self.SSIM_count)]  # SSIMS recorded until median is found
+        self.boiling = [queue.Queue(maxsize=self.SSIM_count),
```
what's the difference between boiling list and SSIM list? (add comments)

benhylak commented on SW/main/Stove.py:
```diff
@@ -5,6 +5,18 @@ def __init__(self):
         self.lower_left = Burner("lower_left")
         self.lower_right = Burner("lower_right")
 
+    def get_burners(self):
```
python's inline array initialization is really powerful.

give it a try:

```
def get_burners(self):
	return [self.upper_left, self.upper_right, self.lower_left, self.lower_right]
```

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -170,71 +164,76 @@ def update(self, frame, stove):
 
                     # ensures that the boiling list is not empty before trying to calculate status
-                    if len(self.boiling[i]) > 0:
+                    if self.boiling[curr_idx].qsize() > 0:
```
let's talk about what you're trying to accomplish here. Might be a better way.

benhylak replied:
using state machine.

mariamedi replied:
I see why you would be suggesting a state machine. In python, would it be best to do this by using enums?

benhylak replied:
yep!

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -170,71 +164,76 @@ def update(self, frame, stove):
+                            # remove one element from the queue every time 30 seconds have passed or boiling began
+                            if self.began_boiling[curr_idx] or time.time() - start_time >= 30:
+                                self.boiling[curr_idx].get()
+                curr_idx = curr_idx + 1
```
index won't be necessary if you take suggestion from above.

benhylak commented on SW/Boiling Detection/boil_detect.py:
```diff
@@ -170,71 +164,76 @@ def update(self, frame, stove):
             # Detects boiling on active burners
-            i = 0
+            curr_idx = 0 # index corresponding to current burner
             for burner in burners:
```
```
for burner, burner_ssim, burner_boiling in zip(burners, median_SSIM, boiling):
	# rest of code -- avoids indexing
```

mariamedi replied:
okay wait quick question, for some of these though I need to modify the element stored at a certain index. So, wouldn't I still need an index variable in some cases bc this is creating a local variable? I did include in the zip the variables I wasn't modifying

benhylak replied:
are you actually changing the reference, or are you just mutating the object?  if the latter, you can mutate the object without an index. (like push/pop elements onto the array)

mariamedi replied:
in the places where i use an index, I believe I want to be changing the reference not just mutating the object

mariamedi committed:
modified to include boil state machine

mariamedi committed:
modified to include update state machine

mariamedi committed:
modified variable constant names, checks SSIM score to see if under a certain thresh to recalculate mask

mariamedi commented:
## Changes
- Modified section of code checking boiling to be a state machine
- Modified all of code within update to be a kind of state machine
- To address the problem of mask if pot is moved:
- SSIM score checked after calculated against a MASK_THRESH, if it is less than it, it breaks out of the loop and exits update function
- breaks because to recalculate the mask a new frame is required

bhylak commented:
The last step here is to test it on the physical device.

mariamedi committed:
modified main to include boil_detector class, split camera capture into distorted vs undistorted

mariamedi committed:
added a crop closer to the stove

mariamedi committed:
fixed some issues discovered while testing, still some issues

mariamedi committed:
added minor file change to test if boiling recorded

mariamedi committed:
changes made at Bens house

mariamedi committed:
modified boiling thresh and min temp thresh
